<a href="https://colab.research.google.com/github/Syomara/CodeAlpha_Data-Analytics-Intership/blob/main/Task_2/Task_2_CodeAlpha_Intership.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **CodeAlpha Intership - Data Analytics**
## Task 2:

*English:*

Objectives:
* Perform Exploratory Data Analysis (EDA) on the database compiled in Task 1
* Evaluate the need for data transformation and data cleaning to remove confusing data.

*Español:*

Objetivos:
* Realizar un Análisis Exploratorio de Datos (EDA) en la base armada en el Task 1.
* Evaluar la necesidad de transformación de datos y limpieza de datos que aporten confusión en los datos.

In [ ]:
#Import the database from GitHub/Importar la base de dato desde GitHub
import pandas as pd

url = "https://raw.githubusercontent.com/Syomara/CodeAlpha_Data-Analytics-Intership/refs/heads/main/Task_1/escaramuza_libros.csv"
df = pd.read_csv(url)

print(df.shape)

(4319, 6)


In [ ]:
#Evaluate df size/Evaluamos la cantidad de datos
df.size

25914

*English:*

The database has 4319 rows and 6 columns, with a total of 25914 items.

We will now begin by evaluating the data type in each column to determine how to proceed.

*Español:*

La base de datos cuenta con 4319 filas y 6 columnas, con un total de 25914 datos.

Comenzaremos ahora evaluando el tipo de dato en cada una de las columnas para evaluar como proseguir.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4319 entries, 0 to 4318
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   titulo          4319 non-null   object
 1   autor           4311 non-null   object
 2   precio          4319 non-null   object
 3   disponibilidad  4319 non-null   object
 4   link            4319 non-null   object
 5   categoria       4319 non-null   object
dtypes: object(6)
memory usage: 202.6+ KB


*English:*

Several relevant pieces of information can be gleaned from this analysis:
* The Author column has missing data.
* The Price column contains text data and should contain numbers.
* The Availability column contains text data and should contain numbers.

Furthermore, I consider it important to assess the existence of duplicates.

We will begin by analyzing the first few rows of the dataset to better understand the issue:

*Español:*

Podemos sacar varias informaciones relevantes de este análisis:
* En la columna de Autor faltan datos
* La columna precio tiene datos de texto y debería ser números
* La columna disponibildad tiene datos de texto y podría ser números

Por otro lado, considero importante evaluar la existencia de duplicados.

Comenzaremos con un análisis de las primeras filas del dataset para entender mejor a que nos enfrentamos:

In [ ]:
df.head()

,titulo,autor,precio,disponibilidad,link,categoria
0,BLOCK IMANTADO 10X20 PRECISO COSAS,VARIOS AUTORES,"UYU 250,75",Disponible,/p/block-imantado-10x20-preciso-cosas/179654/2...,Arte
1,BLOCK IMANTADO 10X10 COLORBLOCK,VARIOS AUTORES,"UYU 250,75",Disponible,/p/block-imantado-10x10-colorblock/179653/283025,Arte
2,EL CAMINO DEL ARTISTA,JULIA CAMERON,"UYU 1.181,50",Disponible,/p/el-camino-del-artista/160123/111268,Arte
3,TECNICAS DE DIBUJO CON PLUMA Y TINTA,DAVID MORALES,UYU 935,Disponible,/p/tecnicas-de-dibujo-con-pluma-y-tinta/178740...,Arte
4,PINTURA DE PAISAJES,MITCHELL ALBALA,UYU 1.105,Disponible,/p/pintura-de-paisajes/173843/277710,Arte


*English:*

We will proceed with the aforementioned points one by one, thus eliminating variables:

*Español:*

Vamos a proceder con los puntos antes mencionados de a uno así vamos eliminando variables:

### Author Column/Columna de autor:

*English:*

We will assess the amount of missing data to determine how to proceed:

*Español:*

Realizaremos una evaluación de la cantidad de datos faltantes para ver como proceder:

In [ ]:
faltantes = df["autor"].isnull().sum() #Amount of missing data/Cantidad de datos faltantes
total = len(df) #Amount of rows/Cantidad de filas del df
porcentaje = (faltantes / total) * 100 #Percentage of missing data/Porcentaje de datos que faltan

print(f"Datos faltantes en 'autor': {faltantes}")
print(f"Total de filas: {total}")
print(f"Porcentaje de faltantes: {porcentaje:.2f}%")

Datos faltantes en 'autor': 8
Total de filas: 4319
Porcentaje de faltantes: 0.19%


*English:*

The percentage of missing data is very low, less than 1%, which at first glance shouldn't affect the overall analysis of the database.

However, I'm going to review those rows to see if there's any useful information before discarding them.

*Español:*

El porcentaje de datos faltantes es muy bajo, menos del 1% lo cual a simple vista no afectará el análisis global del df.

Igualmente, voy a visualizar esas filas para evaluar si hay alguna información que pueda ser útil antes de descartarlos.

In [ ]:
df[df["autor"].isnull()]

,titulo,autor,precio,disponibilidad,link,categoria
402,BAUDEVIN HELLO SPIRALE,NaN,"UYU 671,50",Sin Stock,/p/baudevin-hello-spirale/110618/114695,Arte
437,CUBISM AND CULTURE,NaN,"UYU 514,25",Sin Stock,/p/cubism-and-culture/90707/93799,Arte
723,I CHING,NaN,"UYU 586,50",Disponible,/p/i-ching/112605/116750,Ciencias sociales
860,ABANICOS,NaN,UYU 833,Disponible,/p/abanicos/172050/275904,Ciencias sociales
1554,RECETAS PARA PARRILLAS Y ASADORES,NaN,"UYU 178,50",Disponible,/p/recetas-para-parrillas-y-asadores/71278/73130,Gastronomía
1791,50 RECETAS CON CARNES,NaN,UYU 170,Disponible,/p/50-recetas-con-carnes/162487/74435,Gastronomía
1806,MARRUECOS. LA COCINA MEDITERRANEA,NaN,UYU 238,Disponible,/p/marruecos-la-cocina-mediterranea/66439/73407,Gastronomía
2098,ADIVINA CUÁNTO TE QUIERO (MEDIANO CARTONÉ),NaN,"UYU 671,50",Disponible,/p/adivina-cuanto-te-quiero-mediano-cartone-/1...,Infantiles


*English:*

It's clear that these books don't have the same name and belong to different categories, so I don't see a problem with deleting them.

Therefore, we proceed to delete these rows from the data file (a new data file is also created to keep the original, just in case):

*Español:*

Se observan que no son libros con el mismo nombre y son de varias Categorías distintas, por esto no veo problema en su eliminación.

Por lo tanto, procedemos a eliminar estas filas del df (se crea un df nuevo igualmente para guardar el original por las dudas):

In [ ]:
df_a = df.dropna(subset=["autor"]).copy()

#We evaluated the information from the df to verify the change./Evaluamos la información del df para verificar el cambio.
df_a.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4311 entries, 0 to 4318
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   titulo          4311 non-null   object
 1   autor           4311 non-null   object
 2   precio          4311 non-null   object
 3   disponibilidad  4311 non-null   object
 4   link            4311 non-null   object
 5   categoria       4311 non-null   object
dtypes: object(6)
memory usage: 235.8+ KB


*English:*

It is observed that all columns have the same amount of data, therefore we move on to the next stage.

*Español:*

Se observa que todas las columnas tienen la misma cantidad de datos, por lo tanto pasamos a la siguiente etapa.

### Price Column/Columna de Precio:

*English:*

As noted above, the price column is currently displayed as text; this column includes the currency in words, which is why it's shown as text.

We will proceed as follows:
* Separating the currency and price into two columns
* Changing the price column to a number.

*Español:*

Como se observó en lineas anteriores, la columna precio está como texto; esta columna incluye la moneda con letras por lo que es debido a esto que se encuentra como texto.

Vamos a proceder de la siguiente manera:
* Separando la moneda y el precio en 2 columnas
* Cambiando la columna precio a número.

In [ ]:
#We check what currency formats are available before separating columns/Vemos qué formatos de moneda hay antes de separar
df_a["precio"].str.split(" ").str[0].unique()

array(['UYU'], dtype=object)

In [ ]:
#Two new columns are created with values ​​from the price column/Se crean 2 columnas nuevas con valores de la columna precio
df_a["moneda"] = df_a["precio"].str.split(" ").str[0] #Currency column with the value before the space/Columna Moneda con el valor antes del espacio
df_a["precio_numero"] = df_a["precio"].str.split(" ").str[1] #precio_numero column with the value after the space/Columna precio_numero con el valor despues del espacio

df_a[["precio", "moneda", "precio_numero"]].head()

,precio,moneda,precio_numero
0,"UYU 250,75",UYU,"250,75"
1,"UYU 250,75",UYU,"250,75"
2,"UYU 1.181,50",UYU,"1.181,50"
3,UYU 935,UYU,935
4,UYU 1.105,UYU,1.105


*English:*

We will now convert the precio_numero column to number format.

In the original df, the price is formatted with a period to separate thousands and a comma to separate decimals. We need to transform it into a format that Python recognizes better, which is: with no thousands separator and a period to separate decimals.

*Español:*

Procedemos ahora a convertir la columna precio_numero en formato número.

En el caso del df original el precio está escrito con formato de punto para separar miles y coma para separar decimales; debemos transformarlo a un formato que Python reconozca mejor, este es: sin nada para separar miles y punto para separar decimales.

In [ ]:
df_a["precio_numero"] = (
    df_a["precio_numero"]
    .str.replace(".", "", regex=False)   #Eliminate the period at thousands/Elimina el punto de los miles
    .str.replace(",", ".", regex=False)  #Converts the decimal comma to a period/Convierte la coma decimal en punto
    .astype(float) #Converts the value to a number/Transforma a número el valor
)

df_a[["precio", "precio_numero"]].head()

,precio,precio_numero
0,"UYU 250,75",250.75
1,"UYU 250,75",250.75
2,"UYU 1.181,50",1181.50
3,UYU 935,935.00
4,UYU 1.105,1105.00


*English:*

The change in the way the data is written is observed; we will now evaluate if that column's format has changed:

*Español:*

Se observa el cambio de forma de escribir los datos, evaluaremos ahora si cambió de formato esa columna:

In [ ]:
df_a.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4311 entries, 0 to 4318
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   titulo          4311 non-null   object 
 1   autor           4311 non-null   object 
 2   precio          4311 non-null   object 
 3   disponibilidad  4311 non-null   object 
 4   link            4311 non-null   object 
 5   categoria       4311 non-null   object 
 6   moneda          4311 non-null   object 
 7   precio_numero   4311 non-null   float64
dtypes: float64(1), object(7)
memory usage: 303.1+ KB


*English:*

The precio_numero column is shown in numeric format; we proceed to delete the price column to clean the data file.

*Español:*

Se observa la columna de precio_numero con formato numérico, procedemos a eliminar la columna precio para limpiar el df.

In [ ]:
df_a = df_a.drop(columns=["precio"])

df_a.head()

,titulo,autor,disponibilidad,link,categoria,moneda,precio_numero
0,BLOCK IMANTADO 10X20 PRECISO COSAS,VARIOS AUTORES,Disponible,/p/block-imantado-10x20-preciso-cosas/179654/2...,Arte,UYU,250.75
1,BLOCK IMANTADO 10X10 COLORBLOCK,VARIOS AUTORES,Disponible,/p/block-imantado-10x10-colorblock/179653/283025,Arte,UYU,250.75
2,EL CAMINO DEL ARTISTA,JULIA CAMERON,Disponible,/p/el-camino-del-artista/160123/111268,Arte,UYU,1181.50
3,TECNICAS DE DIBUJO CON PLUMA Y TINTA,DAVID MORALES,Disponible,/p/tecnicas-de-dibujo-con-pluma-y-tinta/178740...,Arte,UYU,935.00
4,PINTURA DE PAISAJES,MITCHELL ALBALA,Disponible,/p/pintura-de-paisajes/173843/277710,Arte,UYU,1105.00


### Availability column/Columna Disponibilidad:

*English:*

We will evaluate the information in the Availability column, since if it is the number of books we could change it to a numeric format.

*Español:*

Evaluaremos cual es la información de la columna Disponibilidad, ya que si es la cantidad de libros podríamos cambiar a un formato numérico.

In [ ]:
df_a["disponibilidad"].unique()

array(['Disponible', 'Sin Stock'], dtype=object)

*English:*

It is decided to leave the column in this way, since it does not reflect quantities of books but only whether or not it is in stock.

*Español:*

Se decide dejar de esta manera la columna, ya que no son cantidades de libros las reflejadas sino sólo si está o no en stock.

### Duplicates/Duplicados:

*English:*

Before finalizing the EDA, we will assess whether there are duplicates in the original database to remove them and simplify the analysis.

We will evaluate repetitions by title, author, and price; this way we ensure that we avoid, for example, the same book with different publishers and prices.

*Español:*

Antes de finalizar el EDA, evaluaremos si existen duplicados en el df original para eliminarlos y simplificar el análisis.

Evaluaremos repeticiones por título, autor y precio; de esta forma nos aseguramos que evitar por ejemplo, el mismo libro con distintas editoriales que tengan precios distintos.

In [ ]:
duplicados = df_a[df_a.duplicated(subset=["titulo", "autor", "precio_numero"], keep="first")]
duplicados.sort_values("titulo")

,titulo,autor,disponibilidad,link,categoria,moneda,precio_numero
3905,A VECES TE SIENTES TAN SOLO QUE TIENE SENTIDO,CHARLES BUKOWSKI,Disponible,/p/a-veces-te-sientes-tan-solo-que-tiene-senti...,Poesía,UYU,1020.0
1271,"ACTUA COMO UN LIDER, PIENSA COMO UN LIDER",HERMINIA IBARRA,Disponible,/p/actua-como-un-lider-piensa-como-un-lider/16...,Economía,UYU,918.0
756,AFINIDADES CONECTIVAS. SOCIOLOGÍA DE LA CULTUR...,VINCENZO SUSCA,Disponible,/p/afinidades-conectivas-sociologia-de-la-cult...,Ciencias sociales,UYU,952.0
3490,ALERTA: ELECCIONES,LEANDRO DELGADO,Disponible,/p/alerta-elecciones/151393/261041,Literatura,UYU,586.5
3898,AMOR ES UN PERRO DEL INFIERNO. POEMAS 1974-1977,"BUKOWSKI, CHARLES",Disponible,/p/amor-es-un-perro-del-infierno-poemas-1974-1...,Poesía,UYU,1266.5
...,...,...,...,...,...,...,...
3140,"VALS DE LA BRUJA, EL (EDICION ESPECIAL)",BELÉN MARTÍNEZ,Disponible,/p/vals-de-la-bruja-el-edicion-especial-/16283...,Juveniles,UYU,1147.5
3342,VIERNES CON LOBOS,MARIO DELGADO APARAÍN,Disponible,/p/viernes-con-lobos/104898/108649,Literatura,UYU,697.0
2652,WIGETTA Y LOS GUSANOS GUASONES,VEGETTA 777,Disponible,/p/wigetta-y-los-gusanos-guasones/82891/85226,Juveniles,UYU,722.5
3392,Y LÍBRANOS DEL MAL,SANTIAGO RONCAGLIOLO,Disponible,/p/y-libranos-del-mal/113294/117467,Literatura,UYU,926.5


*English:*

In most cases, we see that the book is repeated within the same category, so it doesn't contribute to subsequent analyses.

We will evaluate the percentage of duplicates within the total data set:

*Español:*

En la mayoría de los casos vemos que el libro se repite dentro de la misma categoría por lo que no suma a análisis posteriores.

Evaluaremos el % de duplicados dentro del total de datos del df:

In [ ]:
total = len(df_a)
porcentaje = ((len(duplicados)) / total) * 100

print(f"Duplicados: {len(duplicados)}")
print(f"Total de filas: {total}")
print(f"Porcentaje de duplicados: {porcentaje:.2f}%")

Duplicados: 157
Total de filas: 4311
Porcentaje de duplicados: 3.64%


*English:*

The decision to remove them is made for two reasons:
* The percentage of duplicates is small (less than 5%)
* Keeping these duplicates doesn't contribute to my analysis because they repeat data that clutters subsequent analyses.

We proceed to remove them:

*Español:*

Se decide eliminar por 2 motivos:
* El % de duplicados es pequeño (menor al 5%)
* Mantener estos duplicados no suman en mi análisis ya que repiten datos que ensucian los análisis posteriores.

Procedemos a eliminarlos:

In [ ]:
df_a = df_a.drop_duplicates(subset=["titulo", "autor", "precio_numero"], keep="first").copy()
df_a.shape

(4154, 7)

*English:*

We re-evaluate the duplicates:

*Español:*

Volvemos a evaluar los duplicados:

In [ ]:
df_a.duplicated(subset=["titulo", "autor", "precio_numero"]).sum()

np.int64(0)

*English:*

It correctly gave 0, verifying the correctness of the df.

*Español:*

De forma correcta dió 0, verificando la corrección del df.

### Evaluation of Formats and Writing at Author Column:/Evaluación de Formatos y escritura en Columna Autor:

*English:*

The format in which the Authors are written is evaluated, and whether there is any extra space between the first and last name.

*Español:*

Se procede a evaluar el formato en que están escritos los Autores y si existiera algún espacio de más entre nombre y apellido.

In [ ]:
print(f"Autores únicos: {df_a['autor'].nunique()}")
df_a["autor"].unique()[:30]  #First 30, to avoid overloading the screen/Primeros 30, para no saturar la pantalla

Autores únicos: 2610


array(['VARIOS AUTORES', 'JULIA CAMERON', 'DAVID MORALES',
       'MITCHELL ALBALA', 'VERA KAY', 'SERGIO GUINOT',
       'RICHARD WOLFRIK GALLAND', 'NAOMI LORD', 'MARY WOODIN',
       'AMY DEMPSEY', 'MARC PETITJEAN', 'YORIFUJI, BUNPEI', 'AI WEIWEI',
       'EMMA SANGUINETTI', 'LAURA REVUELTA', 'ROSINA  GIL',
       'NACHMANOVITCH STEPHEN', 'NOBUYOSHI ARAKI', 'WAGER, LAUREN',
       'GIULIA MANZI', 'WALTER BENJAMIN', 'CHARLENE  LETENNEUR',
       'SABATIER, ROBERT', 'TILO SCHNEIDER', 'VIKTORIJA SEMJONOVA',
       'ENRIQUE  ETIEVAN', 'AUSTIN KLEON', 'HALLER, KAREN',
       'HENRI CARTIER-BRESSON', 'GEMMA  KOOMEN'], dtype=object)

*English:*

Two problems were observed:
1. Authors with one or more spaces between their first and last name, which can unnecessarily generate duplicate authors.

2. Formats written as Last Name, First Name and as First Name, Last Name.

The second problem is too difficult to solve, since attempting to resolve it automatically could produce errors in the results, providing incorrect information. Therefore, we will proceed to address the first error:

*Español:*

Se observan 2 problemas:
1. Autores con 1 o más espacios entre nombre y apellido, esto puede generar duplicados de autores de forma innecesaria.
2. Formatos escritos como Apellido, Nombre y como Nombre Apellido.

El segundo problema es demasiado dificil de solucionar, ya que si intentamos solucionarlo de forma automática podemos producir errores en los resultados de forma que nos entregue información incorrecta. Procedemos entonces a solucionar el 1er error:

In [ ]:
df_a["autor"] = df_a["autor"].str.replace(r"\s+", " ", regex=True).str.strip()

*English:*

We re-evaluated the list of unique authors to assess whether this change helped reduce potential duplicates:

*Español:*

Evaluamos de nuevo el listado de autores unicos para evaluar si este cambio ayudó a disminuir posibles duplicados:

In [ ]:
print(f"Autores únicos: {df_a['autor'].nunique()}")
df_a["autor"].unique()[:30]  ##First 30, to avoid overloading the screen/Primeros 30, para no saturar la pantalla

Autores únicos: 2598


array(['VARIOS AUTORES', 'JULIA CAMERON', 'DAVID MORALES',
       'MITCHELL ALBALA', 'VERA KAY', 'SERGIO GUINOT',
       'RICHARD WOLFRIK GALLAND', 'NAOMI LORD', 'MARY WOODIN',
       'AMY DEMPSEY', 'MARC PETITJEAN', 'YORIFUJI, BUNPEI', 'AI WEIWEI',
       'EMMA SANGUINETTI', 'LAURA REVUELTA', 'ROSINA GIL',
       'NACHMANOVITCH STEPHEN', 'NOBUYOSHI ARAKI', 'WAGER, LAUREN',
       'GIULIA MANZI', 'WALTER BENJAMIN', 'CHARLENE LETENNEUR',
       'SABATIER, ROBERT', 'TILO SCHNEIDER', 'VIKTORIJA SEMJONOVA',
       'ENRIQUE ETIEVAN', 'AUSTIN KLEON', 'HALLER, KAREN',
       'HENRI CARTIER-BRESSON', 'GEMMA KOOMEN'], dtype=object)

*English:*

It is observed that there were no changes regarding the number of authors, so the double spacing did not affect the previous count.

*Español:*

Se observa que no hubo cambios al respecto de la cantidad de autores, por lo que el doble espacio no afectaba el conteo anterior.

*English:*

To avoid losing what we've done, we save it on GitHub as a csv file so we can continue using it in Task 3.

*Español:*

Para no perder lo echo, lo guardamos en GitHub como un archivo csv y así podemos continuar usandolo en el Task 3.

In [ ]:
#We deleted it in case there was a file with that name/Borramos por si existe algún archivo con se nombre
!rm -rf CodeAlpha_Data-Analytics-Intership
#The repository to be used is cloned/Se clona el repositorio a usar
!git clone https://github.com/Syomara/CodeAlpha_Data-Analytics-Intership.git

Cloning into 'CodeAlpha_Data-Analytics-Intership'...
remote: Enumerating objects: 67, done.
remote: Counting objects: 100% (67/67), done.
remote: Compressing objects: 100% (63/63), done.
remote: Total 67 (delta 20), reused 4 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (67/67), 456.50 KiB | 4.15 MiB/s, done.
Resolving deltas: 100% (20/20), done.


In [ ]:
#We evaluate what information is brought from the Repository/Evaluamos que información se trae desde el Repositorio
!ls /content/CodeAlpha_Data-Analytics-Intership/

 Carta_de_confirmación.pdf		    README.md   Task_2
'Data_Analytics_Tasks_&_Instructions.pdf'   Task_1


*English:*

As we observed in the previous line, the Repository already contains the Task 2 folder, so we can save directly there:

*Español:*

Como observamos en la linea anterior, el Repositorio ya contiene la carpeta del Task 2 por lo que podemos directamente guardar ahí:

In [ ]:
df_a.to_csv("/content/CodeAlpha_Data-Analytics-Intership/Task_2/escaramuza_libros_limpio.csv", index=False)

In [ ]:
#We uploaded the file to GitHub/Subimos el archivo a GitHub
from google.colab import userdata
token = userdata.get('GitHub_token')

%cd /content/CodeAlpha_Data-Analytics-Intership/Task_2/
!git add escaramuza_libros_limpio.csv
!git commit -m "Agrego dataset de libros con EDA realizado"
!git config user.email "syomarabazz@hotmail.com"
!git config user.name "Syomara"
!git push https://{token}@github.com/Syomara/CodeAlpha_Data-Analytics-Intership.git

/content/CodeAlpha_Data-Analytics-Intership/Task_2
[main 498c7ac] Agrego dataset de libros con EDA realizado
 1 file changed, 4160 insertions(+)
 create mode 100644 Task_2/escaramuza_libros_limpio.csv
Enumerating objects: 6, done.
Counting objects: 100% (6/6), done.
Delta compression using up to 2 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 171.25 KiB | 4.76 MiB/s, done.
Total 4 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/Syomara/CodeAlpha_Data-Analytics-Intership.git
   600e2cb..498c7ac  main -> main
